[3팀 여진석] Chapter 2-1. 클린코드와 리팩토링

realstone2 opened the pull request:
[배포링크](https://realstone2.github.io/front_6th_chapter2-1/)
## 과제 체크포인트

### 기본과제

- [x] 코드가 Prettier를 통해 일관된 포맷팅이 적용되어 있는가?
- [x] 적절한 줄바꿈과 주석을 사용하여 코드의 논리적 단위를 명확히 구분했는가?
- [x] 변수명과 함수명이 그 역할을 명확히 나타내며, 일관된 네이밍 규칙을 따르는가?
- [x] 매직 넘버와 문자열을 의미 있는 상수로 추출했는가?
- [x] 중복 코드를 제거하고 재사용 가능한 형태로 리팩토링했는가?
- [x] 함수가 단일 책임 원칙을 따르며, 한 가지 작업만 수행하는가?
- [x] 조건문과 반복문이 간결하고 명확한가? 복잡한 조건을 함수로 추출했는가?
- [x] 코드의 배치가 의존성과 실행 흐름에 따라 논리적으로 구성되어 있는가?
- [x] 연관된 코드를 의미 있는 함수나 모듈로 그룹화했는가?
- [x] ES6+ 문법을 활용하여 코드를 더 간결하고 명확하게 작성했는가?
- [x] 전역 상태와 부수 효과(side effects)를 최소화했는가?
- [x] 에러 처리와 예외 상황을 명확히 고려하고 처리했는가?
- [x] 코드 자체가 자기 문서화되어 있어, 주석 없이도 의도를 파악할 수 있는가?
- [x] 비즈니스 로직과 UI 로직이 적절히 분리되어 있는가?
- [x] 코드의 각 부분이 테스트 가능하도록 구조화되어 있는가?
- [x] 성능 개선을 위해 불필요한 연산이나 렌더링을 제거했는가?
- [x] 새로운 기능 추가나 변경이 기존 코드에 미치는 영향을 최소화했는가?
- [] 코드 리뷰를 통해 다른 개발자들의 피드백을 반영하고 개선했는가?
- [x] (핵심!) 리팩토링 시 기존 기능을 그대로 유지하면서 점진적으로 개선했는가?

### 심화과제

- [x] 변경한 구조와 코드가 기존의 코드보다 가독성이 높고 이해하기 쉬운가?
- [x] 변경한 구조와 코드가 기존의 코드보다 기능을 수정하거나 확장하기에 용이한가?
- [x] 변경한 구조와 코드가 기존의 코드보다 테스트를 하기에 더 용이한가?
- [x] 변경한 구조와 코드가 기존의 모든 기능은 그대로 유지했는가?
- [x] (핵심!) 변경한 구조와 코드를 새로운 한번에 새로만들지 않고 점진적으로 개선했는가?

## 과제 셀프회고
이번에는 직접 코딩하지 않고 AI를 활용해서 방향성만 지시해서 과제를 완료하려고 해보았습니다.
한번에 정리해줘! 하는식의 방식으로는 절대 원하는 방향성으로 해주지 않았습니다.
그래서 이번에는 AI를 어떻게 사용해야지 내가 원하는 결과물을 얻을 수 있을까에 대해서 많은 고민을 했던 과제였습니다.


<!-- 과제에 대한 회고를 작성해주세요 -->

### 과제를 하면서 내가 제일 신경 쓴 부분은 무엇인가요?
제가 원하는 리팩토링 구조에 맞춰서 AI가 잘 나눌 수 있도록 프롬프트, prd 작성하는 것을 가장 신경쓴 것 같습니다.
basic 과제에서는 view => model + viewmodel => event 순서로 구조를 나눠가도록 프롬프트를 단계적으로 요청하였습니다.
advanced 과제에서는 준형님이 추천해주신 bmad-method를 사용하여 prd 문서를 만들어서 원하는 구조를 미리 잡고 basic과제의 구조와 동일하게 구성하였습니다.
이런 구조 자체에 대해서 신경을 많이 쓴 것 같습니다.


### 과제를 다시 해보면 더 잘 할 수 있었겠다 아쉬운 점이 있다면 무엇인가요?
레거시 코드의 액션들을 모두 viewmodel쪽에 한번에 적용하다보니 너무 크게 추상화를 하게 되어 아쉽습니다.
model을 jotai로 선택하지 않고 reducer action 기반으로 viewmodel을 작성했더라면 좀 더 깔끔하지 않았을까 하는 생각이 있습니다.

테스트코드를 기본과제 내용 기반으로 미리 더 잘 작성되어야되었을 것 같습니다.
테스트코드가 의미가 없을정도로 통과했으나 기능이 누락되는 일이 있었습니다. 테스트코드의 중요성을 더 깨닫게된 것 같습니다.




### 리뷰 받고 싶은 내용이나 궁금한 것에 대한 질문 편하게 남겨주세요 :)
1. 이런 리팩토링을 진행할 때 테스트코드가 제일 중요하다고 느낀 것 같습니다.
AI한테 기본 과제 테스트코드 내용을 그대로 반영해서 React에서 사용하는 코드 기반으로 테스트코드를 만들어달라고 했는데, 기능 자체를 누락해서 만들어주는 일들이 있었습니다.
테스트 코드에서 이런일이 발생해서는 안될텐데, 테스트코드를 AI에게 작성시키는 것에 대해서 어떻게 생각하시는지 궁금합니다!

2. 이렇게 큰 파일 기반의 리팩토링할 때 AI를 사용했더니 비용이 너무 많이 나가게되는 것 같습니다. 이럴 때 token비용을 절약할 수 있는 노하우가 있을까요?

yeojinseok1 committed:
Refactor code structure and remove redundant changes

yeojinseok1 committed:
Implement feature enhancements and bug fixes in the user authentication module

yeojinseok1 committed:
📜 리팩토링 계획 문서 추가

`main.basic.js`의 클린코드 리팩토링을 위한 개발 계획 및 현재 상황 분석을 포함한 `rule.md` 파일을 새로 추가했습니다. 이 문서에는 리팩토링 목표, 제약사항, 현재 코드 상태, 개발 전략 및 성공 지표가 상세히 설명되어 있습니다.

yeojinseok1 committed:
✨ 새로운 컴포넌트 및 스토어 관련 기능 추가

`Component` 클래스를 포함한 새로운 컴포넌트 구조를 추가하였으며, `createObserver`, `createStorage`, `createStore`, `shallowEquals`와 같은 스토어 관련 유틸리티를 구현했습니다. 이로써 상태 관리 및 데이터 저장 기능을 강화하였습니다.

yeojinseok1 committed:
✨ Prettier 및 ESLint 설정 추가

`.prettierignore`, `.prettierrc`, `eslint.config.js` 파일을 새로 추가하여 코드 포맷팅 및 린팅 규칙을 설정했습니다. 또한, HTML 파일의 구조를 개선하고, `package.json`에 린트 및 포맷 관련 스크립트를 추가했습니다. 이를 통해 코드 품질을 향상시키고 일관성을 유지할 수 있도록 하였습니다.

yeojinseok1 committed:
fix: resolve all lint errors in main.original.js - replace var with const/let declarations - fix variable reassignment issues - remove unused variables - add meaningful comments for empty blocks

yeojinseok1 committed:
feat: 1차 코드 정리

yeojinseok1 committed:
refactor: 상수 및 전역 상태 변수에 대한 주석 추가와 코드 정리

상수 정의 및 전역 상태 변수에 대한 설명을 추가하여 코드 가독성을 향상시켰습니다. 각 상수의 역할과 사용 목적을 명확히 하여 유지보수성을 높였습니다. 또한, 애플리케이션 초기화 및 주요 함수들에 대한 주석을 추가하여 코드 흐름을 이해하기 쉽게 개선하였습니다.

yeojinseok1 committed:
refactor: html 리터럴 반환하는 함수 분리

yeojinseok1 committed:
refactor: 컴포넌트 element return 형태로 리팩토링

yeojinseok1 committed:
refactor: 상수 분리

yeojinseok1 committed:
refactor: 고차함수로 리팩토링 및 순수함수 파일 분리

yeojinseok1 committed:
refactor: 불필요 파일 제거 및 ts 마이그레이션

yeojinseok1 committed:
feat: ProductState Reducer
`productState.ts` 파일을 새로 생성하여 전역 상태 관리 기능을 구현했습니다. 이를 통해 상품 데이터의 상태를 효율적으로 관리할 수 있게 되었습니다. 또한, 기존 코드에서 상품 관련 로직을 상태 관리로 통합하여 코드의 가독성과 유지보수성을 향상시켰습니다.

yeojinseok1 committed:
feat: 장바구니 상태 관리 기능 추가

`cartState.ts` 파일을 새로 생성하여 장바구니 도메인 전역 상태 관리 기능을 구현했습니다. 이를 통해 장바구니 아이템의 추가, 삭제, 수량 업데이트 및 총액 계산을 효율적으로 처리할 수 있게 되었습니다. 또한, `main.basic.ts` 파일에서 장바구니 상태를 초기화하고 업데이트하는 로직을 추가하여 코드의 가독성과 유지보수성을 향상시켰습니다.

yeojinseok1 committed:
feat: 컴포넌트 계층 분리

yeojinseok1 committed:
feat: UI 상태 관리 기능 추가

`uiState.ts` 파일을 새로 생성하여 UI 도메인 전역 상태 관리 기능을 구현했습니다. 이를 통해 모달, 헤더, 알림 등 다양한 UI 상태를 효율적으로 관리할 수 있게 되었습니다. 또한, `main.basic.ts` 파일에서 UI 상태 초기화 및 업데이트 로직을 추가하여 코드의 가독성과 유지보수성을 향상시켰습니다.

yeojinseok1 committed:
feat: 포인트 상태 관리 기능 추가

`pointsState.ts` 파일을 새로 생성하여 포인트 도메인 전역 상태 관리 기능을 구현했습니다. 이를 통해 포인트 계산, 적립 내역 관리 및 포인트 표시 설정을 효율적으로 처리할 수 있게 되었습니다. 또한, `main.basic.ts` 파일에서 포인트 상태 초기화 및 업데이트 로직을 추가하여 코드의 가독성과 유지보수성을 향상시켰습니다.

yeojinseok1 committed:
feat: 이벤트 관리 시스템 추가

`eventManager.ts` 파일을 새로 생성하여 이벤트 위임 기반의 전역 이벤트 관리 시스템을 구현했습니다. 이를 통해 각 컴포넌트에서 전역적으로 하나의 이벤트만 등록하여 효율적으로 이벤트를 관리할 수 있게 되었습니다. 코드의 가독성과 유지보수성을 향상시키는 데 기여합니다.

yeojinseok1 committed:
feat: 매뉴얼 섹션 컴포넌트 추가 및 이벤트 리스너 통합

`ManualSection.ts` 파일을 새로 생성하여 매뉴얼 토글 및 오버레이 기능을 통합했습니다. `main.basic.ts`에서 이벤트 리스너를 설정하는 로직을 간소화하고, `App.ts`에서 매뉴얼 섹션을 추가하여 코드의 가독성과 유지보수성을 향상시켰습니다. 또한, 기존의 매뉴얼 관련 코드 일부를 제거하여 구조를 개선했습니다.

yeojinseok1 committed:
feat: 장바구니 관련 이벤트 핸들러 및 컴포넌트 분리

장바구니 추가 버튼과 장바구니 아이템 클릭 이벤트를 각각 `AddToCartButton` 및 `CartDisplay` 컴포넌트로 분리하여 코드의 가독성과 유지보수성을 향상시켰습니다. 또한, 장바구니 계산 및 UI 업데이트를 담당하는 `handleCalculateCartStuff` 함수를 새로 생성하여 장바구니 관련 로직을 정리했습니다. 이로 인해 이벤트 관리가 보다 효율적으로 이루어질 수 있게 되었습니다.

yeojinseok1 committed:
feat: 상품 선택 관련 컴포넌트 및 이벤트 핸들러 분리

`ProductSelector` 컴포넌트를 새로 생성하여 상품 선택 드롭다운의 로직을 분리하였고, 상품 선택 옵션 업데이트 기능을 `productEventHandlers.ts`로 이동시켜 코드의 가독성과 유지보수성을 향상시켰습니다. 또한, `main.basic.ts`에서 상품 선택 관련 주석을 업데이트하여 코드의 명확성을 높였습니다.

yeojinseok1 committed:
feat: 상품 타이머 이벤트 핸들러 분리 및 통합

상품 관련 타이머 이벤트 핸들러를 `productTimerHandlers.ts` 파일로 분리하여 코드의 가독성과 유지보수성을 향상시켰습니다. `main.basic.ts`에서 타이머 관련 로직을 제거하고, `ProductSelector` 컴포넌트에서 타이머를 시작하도록 수정하였습니다. 이를 통해 타이머 기능이 보다 명확하게 관리될 수 있게 되었습니다.

yeojinseok1 committed:
feat: 재고 정보 업데이트 로직 분리 및 주석 정리

재고 정보 업데이트 로직을 `stockEventHandlers.ts`로 분리하여 코드의 가독성과 유지보수성을 향상시켰습니다. `main.basic.ts`에서 재고 관련 주석을 업데이트하고, `StockInformation` 컴포넌트에서 초기 재고 정보를 설정하도록 수정하였습니다. 이를 통해 재고 관리 기능이 보다 명확하게 관리될 수 있게 되었습니다.

yeojinseok1 committed:
feat: 장바구니 가격 업데이트 로직 분리 및 통합

장바구니 내 상품 가격 업데이트 관련 로직을 `cartPriceHandlers.ts`로 분리하여 코드의 가독성과 유지보수성을 향상시켰습니다. `main.basic.ts`에서 가격 업데이트 관련 주석을 수정하고, 상품 타이머 이벤트 핸들러에서 `doUpdatePricesInCart` 함수를 호출하도록 변경하여 장바구니 가격 업데이트 기능을 명확하게 관리할 수 있게 되었습니다.

yeojinseok1 committed:
feat: 장바구니 계산 로직을 orderSummaryHandlers.ts로 이동 및 관련 주석 업데이트

장바구니 계산 및 UI 업데이트 로직을 `orderSummaryHandlers.ts`로 분리하여 코드의 가독성과 유지보수성을 향상시켰습니다. `main.basic.ts`와 여러 컴포넌트에서 `handleCalculateCartStuff` 함수를 호출하도록 수정하여 장바구니 관련 기능을 명확하게 관리할 수 있게 되었습니다.

yeojinseok1 committed:
feat: 전역 상태 변수 및 주석 정리

전역 상태 변수 관련 주석을 제거하고, `main.basic.ts`에서 불필요한 코드와 주석을 정리하여 가독성을 향상시켰습니다. 또한, `useCartState`, `usePointsState` 등 새로운 상태 관리 훅을 도입하여 코드 구조를 개선했습니다.

realstone2 committed:
Merge branch 'main' of github.com:realstone2/front_6th_chapter2-1

realstone2 committed:
refactor: reorganize folder structure to basic/advanced pattern

realstone2 committed:
feat: React 19.1.1 버전 추가

`package.json` 및 `pnpm-lock.yaml` 파일에 React 19.1.1 버전을 추가하여 의존성을 업데이트했습니다.

realstone2 committed:
refactor: import 경로 수정 및 정리

`main.basic.ts` 파일에서 import 경로를 상대 경로로 수정하고, 코드의 가독성을 높이기 위해 불필요한 주석을 정리했습니다. 이를 통해 코드 구조가 더욱 명확해졌습니다.

realstone2 committed:
feat(package): @testing-library/react 패키지 추가

realstone2 committed:
feat(docs): React 마이그레이션 계획 문서 추가

- React 기반 애플리케이션으로의 마이그레이션 전략 수립
- 환경 점검, Vite 설정, 최소 작동 앱 생성 등 단계별 작업 목록 포함
- 기존 코드 분석 및 테스트 전략 제시

realstone2 committed:
feat: Add React Migration PRD and remove Test Code Development PRD

- Introduced a comprehensive Product Requirements Document (PRD) for the React migration of the Hanghae shopping mall project, detailing project vision, goals, technical requirements, and success metrics.
- Removed the Test Code Development PRD as part of the project restructuring.

realstone2 committed:
feat(docs): Jotai 기반 MVVM 아키텍처 설계 및 상태 관리 전략 추가

- Jotai를 활용한 상태 관리 전략을 MVVM 패턴으로 재구성
- 새로운 Jotai atom 및 ViewModel 훅 추가
- 기존 상태 관리 방식에서 Jotai 기반으로 전환

realstone2 committed:
feat: Implement basic structure for advanced shopping cart application

- Add App component as the main entry point for the application.
- Create advanced test suite for shopping cart functionality.
- Set up main entry file to render the App component.
- Configure Vite to use React plugin for proper JSX handling.

realstone2 committed:
feat(react): Phase 1 View 컴포넌트 구조 구축 완료

- 기본 레이아웃 컴포넌트 구현 (App, Header, GridContainer, LeftColumn)
- 장바구니 관련 View 컴포넌트 구현 (CartDisplay, AddToCartButton, SelectorContainer, StockInformation, ManualSection)
- 주문 관련 View 컴포넌트 구현 (OrderSummary)
- 도움말 관련 View 컴포넌트 구현 (ManualColumn, ManualOverlay, ManualToggle)
- 상품 관련 View 컴포넌트 구현 (ProductSelector)
- 모든 컴포넌트에 TODO 주석으로 다음 Phase 구현 계획 명시
- 기존 HTML 구조를 React 컴포넌트로 변환하여 UI 레이아웃 완성
- View 우선 개발 전략으로 UI 구조를 먼저 완성한 후 비즈니스 로직 추가 예정

realstone2 committed:
docs(prd): View 우선 개발 전략 상세화 및 Phase 1 완료 상태 업데이트

- View 우선 개발의 실제 구현 특징 상세 설명 추가
- 14개 View 컴포넌트 완료 상태 반영
- View 우선 개발의 성공 지표 명시
- Phase 1 체크리스트 완료 상태 업데이트
- Phase 2에서 ViewModel 전환 계획 구체화
- 기존 HTML 구조 보존, TODO 주석 체계, 이벤트 핸들러 준비 등 실제 적용 사례 추가

realstone2 committed:
feat(models): 장바구니, 주문, 포인트, 상품, UI 모델 추가

- 장바구니, 주문, 포인트, 상품, UI 도메인에 대한 데이터 타입 및 초기 상태 정의
- Jotai를 이용한 상태 관리 atom 추가

realstone2 committed:
feat(cart): 장바구니 모델 및 상태 추가

- 장바구니 아이템, 할인 정보, 상태 모델 인터페이스 정의
- 초기 장바구니 상태 및 Jotai atom 추가

realstone2 committed:
feat: cart viewModel 구현

realstone2 committed:
feat: order viewmodel 연결

realstone2 committed:
feat: 상품 및 재고 ViewModel 구현

- 상품 도메인에 대한 비즈니스 로직과 상태 관리를 위한 ViewModel 추가
- Jotai를 기반으로 한 상품 및 재고 상태 관리 훅 구현
- 상품 관련 기능 (할인 적용, 재고 조정 등) 및 테스트 케이스 추가
- 상품과 재고 ViewModel 간의 상태 동기화 기능 구현

realstone2 committed:
feat: 포인트 ViewModel 및 주문 ViewModel 통합

- 포인트 ViewModel을 추가하여 포인트 적립 관련 비즈니스 로직 구현
- 주문 ViewModel에서 포인트 계산 및 업데이트 기능 통합
- 포인트 ViewModel에 대한 테스트 케이스 추가 및 기존 주문 ViewModel 테스트 수정

realstone2 committed:
feat(react): MVVM 패턴 기반 React 마이그레이션 Phase 1 완료

- Model Layer: 5개 도메인별 Jotai atom 기반 상태 관리 구현
- ViewModel Layer: 8개 ViewModel 훅 구현 (Cart, Order, Product, Points, UI)
- View Layer: 14개 React 컴포넌트 구현 및 ViewModel 연결
- Test Layer: 27개 테스트 케이스 구현 및 통과
- MVVM 패턴: Model-View-ViewModel 완벽 분리 및 상태 동기화 구현
- Jotai Provider: 반응형 상태 관리 및 테스트 환경 구축

Resolves: React 마이그레이션 Phase 1 완료

realstone2 committed:
테스트: StockInformation 컴포넌트의 재고 상태 메시지 정규식 매칭으로 수정 및 재고 수량 업데이트

- 재고 수량을 10에서 20으로 변경
- 재고 상태 메시지의 텍스트 매칭을 정규식으로 수정하여 가독성 향상
- 여러 테스트 케이스에서 재고 수량을 업데이트하고 정규식 매칭 적용

realstone2 committed:
테스트: CartDisplay 컴포넌트에 대한 유닛 테스트 추가

- 장바구니 아이템 표시, 수량 조절, 아이템 제거 기능에 대한 테스트 케이스 구현
- 장바구니가 비어있을 때의 메시지 표시 및 할인 적용 시 할인 정보 표시 확인
- 각 테스트 케이스에서 초기 데이터 설정을 위한 테스트 래퍼 구성

realstone2 committed:
feat: GitHub Pages 배포 워크플로우 추가

- main 브랜치에 푸시 시 GitHub Pages로 배포하는 워크플로우 구현
- Node.js 및 pnpm 설정, 의존성 설치, 빌드 및 배포 단계 포함
- peaceiris/actions-gh-pages를 사용하여 배포 자동화

realstone2 committed:
build: GitHub Pages 배포 디렉토리 수정

- 배포 디렉토리를 ./src/advanced/dist에서 ./dist로 변경
- Vite 설정 파일에서 빌드 출력 디렉토리 설정 추가

realstone2 committed:
feat: Cart 및 Order ViewModel 개선 및 테스트 추가

- CartViewModel에서 재고 확인 및 수량 조정 로직 개선
- OrderViewModel에서 주문 상태 자동 업데이트 기능 수정
- PointsViewModel에 화요일 여부 확인 기능 추가
- Cart 및 Order 관련 테스트 케이스 추가 및 수정
- OrderSummary 컴포넌트에서 불필요한 useEffect 제거

realstone2 committed:
feat: 404 페이지 추가 및 index.html 수정

- 404.html 파일을 새로 추가하여 사용자 정의 404 페이지 구현
- index.html에서 스크립트 경로를 수정하여 상대 경로로 변경
- vite.config.js에서 빌드 입력 파일을 index.html로 수정

realstone2 committed:
build(config): 빌드 설정 수정

- base 경로와 copyPublicDir 옵션 추가

realstone2 committed:
build(.gitignore): .gitignore 파일 수정

- .bmad-core, .cursor, .github/chatmodes 항목 추가
- .claude 항목 수정

realstone2 committed:
Remove obsolete chatmode files for BMad Orchestrator, Developer, Product Manager, Product Owner, QA Architect, Scrum Master, and UX Expert agents to streamline the codebase.

tooth-is-silver commented on src/basic/features/cart/SelectorContainer.ts:
```diff
@@ -0,0 +1,31 @@
+ * 선택 컨테이너 컴포넌트
+ * @returns 선택 컨테이너 DOM Element
+ */
+export function SelectorContainer(): HTMLElement {
+  const selectorContainer = document.createElement('div');
+  selectorContainer.className = 'mb-6 pb-6 border-b border-gray-200';
+
+  // 상품 선택기 생성
+  const productSelector = ProductSelector();
+  productSelector.className =
+    'w-full p-3 border border-gray-300 rounded-lg text-base mb-3';
+
+  // 장바구니 추가 버튼 생성
+  const addToCartButton = AddToCartButton();
+  addToCartButton.id = 'add-to-cart';
+
+  // 재고 정보 생성
+  const stockInformation = StockInformation();
+
+  // 자식 요소들을 컨테이너에 추가
+  selectorContainer.appendChild(productSelector);
+  selectorContainer.appendChild(addToCartButton);
+  selectorContainer.appendChild(stockInformation);
+
+  return selectorContainer;
+}
```
와 ...나도 비슷하게 만든거 같은데 더 깔끔하다... 부럽다

tooth-is-silver commented on src/app/useState.ts:
```diff
@@ -0,0 +1,26 @@
+
+type SetStateAction<S> = S | ((prevState: S) => S);
+
+export function useState<S>(
```
`useState`를 직접 만드셨군요 대박 👍

JunilHwang replied:
하지만 쓰이진 않았네요 ㅎㅎ

realstone2 replied:
ㅋㅋㅋㅋㅋㅋㅋㅋ 맞아요 만들긴했는데, 리팩토링과정에서는 투머치인거같아서 안했어요

realstone2 committed:
feat(viewmodel): 상품 ViewModel에 타이머 기능 추가

- 번개세일 및 추천세일 타이머 시작 및 정리 기능 구현
- 타이머 관련 상태 관리 및 핸들러 추가
- 컴포넌트 언마운트 시 타이머 정리 로직 추가

realstone2 committed:
feat(docs): 과제 체크포인트 및 셀프회고 문서 추가

- 기본과제와 심화과제 체크리스트 추가
- 과제 셀프회고 섹션 작성
